File: security/src/pairing.cpp
#include "hearport/security/pairing.h"

#include <algorithm>
#include <string_view>

#include "hearport_spake2_provider.h"

namespace hearport::security {
namespace {

template <typename T>
const std::uint8_t* AsU8(std::span<const T> bytes) noexcept {
  return reinterpret_cast<const std::uint8_t*>(bytes.data());
}

template <typename T>
std::uint8_t* AsU8(std::span<T> bytes) noexcept {
  return reinterpret_cast<std::uint8_t*>(bytes.data());
}

template <std::size_t N>
const std::uint8_t* AsU8(const std::array<std::byte, N>& bytes) noexcept {
  return reinterpret_cast<const std::uint8_t*>(bytes.data());
}

template <std::size_t N>
std::uint8_t* AsU8(std::array<std::byte, N>& bytes) noexcept {
  return reinterpret_cast<std::uint8_t*>(bytes.data());
}

constexpr std::string_view kWindowsIdentity = "HearPort-Windows-v1";
constexpr std::string_view kReceiverIdentity = "HearPort-Receiver-v1";

}  // namespace

bool IsValidPin(std::span<const std::byte> pin) noexcept {
  if (pin.size() != kPairingPinBytes) {
    return false;
  }
  return std::ranges::all_of(pin, [](std::byte value) {
    const auto digit = std::to_integer<unsigned int>(value);
    return digit >= static_cast<unsigned int>('0') &&
           digit <= static_cast<unsigned int>('9');
  });
}

std::optional<Bytes32> PinToScalar(
    std::span<const std::byte> pin) noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (!IsValidPin(pin)) {
    return std::nullopt;
  }
  Bytes32 scalar{};
  if (hearport_spake2_pin_to_scalar(AsU8(pin), pin.size(), AsU8(scalar)) !=
      HEARPORT_SPAKE2_OK) {
    return std::nullopt;
  }
  return scalar;
#else
  (void)pin;
  return std::nullopt;
#endif
}

std::optional<Bytes32> Sha256Spki(
    std::span<const std::byte> der_subject_public_key_info) noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (der_subject_public_key_info.empty()) {
    return std::nullopt;
  }
  Bytes32 digest{};
  if (hearport_spake2_sha256(AsU8(der_subject_public_key_info),
                             der_subject_public_key_info.size(),
                             AsU8(digest)) != HEARPORT_SPAKE2_OK) {
    return std::nullopt;
  }
  return digest;
#else
  (void)der_subject_public_key_info;
  return std::nullopt;
#endif
}

std::vector<std::byte> BuildWindowsIdentity(const Bytes32& spki_sha256) {
  std::vector<std::byte> identity;
  identity.reserve(kWindowsIdentity.size() + 1 + spki_sha256.size());
  for (const char value : kWindowsIdentity) {
    identity.push_back(static_cast<std::byte>(value));
  }
  identity.push_back(std::byte{0});
  identity.insert(identity.end(), spki_sha256.begin(), spki_sha256.end());
  return identity;
}

std::vector<std::byte> BuildReceiverIdentity() {
  std::vector<std::byte> identity;
  identity.reserve(kReceiverIdentity.size());
  for (const char value : kReceiverIdentity) {
    identity.push_back(static_cast<std::byte>(value));
  }
  return identity;
}

std::optional<Bytes32> HmacSha256(std::span<const std::byte> key,
                                  std::span<const std::byte> message) noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (key.empty()) {
    return std::nullopt;
  }
  Bytes32 mac{};
  if (hearport_spake2_hmac_sha256(AsU8(key), key.size(), AsU8(message),
                                  message.size(), AsU8(mac)) !=
      HEARPORT_SPAKE2_OK) {
    return std::nullopt;
  }
  return mac;
#else
  (void)key;
  (void)message;
  return std::nullopt;
#endif
}

bool RandomBytes(std::span<std::byte> output) noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  return hearport_spake2_random(AsU8(output), output.size()) ==
         HEARPORT_SPAKE2_OK;
#else
  (void)output;
  return false;
#endif
}

bool ConstantTimeEqual(std::span<const std::byte> left,
                       std::span<const std::byte> right) noexcept {
  if (left.size() != right.size()) {
    return false;
  }
  volatile std::uint8_t difference = 0;
  for (std::size_t index = 0; index < left.size(); ++index) {
    difference |= static_cast<std::uint8_t>(
        std::to_integer<unsigned int>(left[index]) ^
        std::to_integer<unsigned int>(right[index]));
  }
  return difference == 0;
}

struct Spake2Session::Opaque {};
struct Spake2Result::Opaque {};

Spake2Session::~Spake2Session() {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (session_ != nullptr) {
    hearport_spake2_session_free(
        reinterpret_cast<hearport_spake2_session*>(session_));
  }
#endif
}

Spake2Session::Spake2Session(Spake2Session&& other) noexcept
    : session_(other.session_), public_point_(other.public_point_) {
  other.session_ = nullptr;
}

Spake2Session& Spake2Session::operator=(Spake2Session&& other) noexcept {
  if (this == &other) {
    return *this;
  }
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (session_ != nullptr) {
    hearport_spake2_session_free(
        reinterpret_cast<hearport_spake2_session*>(session_));
  }
#endif
  session_ = other.session_;
  public_point_ = other.public_point_;
  other.session_ = nullptr;
  return *this;
}

bool Spake2Session::ProviderAvailable() noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  return true;
#else
  return false;
#endif
}

std::optional<Spake2Session> Spake2Session::Begin(
    Spake2Role role, const Bytes32& scalar,
    std::span<const std::byte> identity_a,
    std::span<const std::byte> identity_b) noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (identity_a.size() > 65536 || identity_b.size() > 65536) {
    return std::nullopt;
  }
  Bytes65 public_point{};
  hearport_spake2_session* session = nullptr;
  const auto result = hearport_spake2_begin(
      static_cast<std::uint8_t>(role), AsU8(scalar), AsU8(identity_a),
      identity_a.size(), AsU8(identity_b), identity_b.size(),
      AsU8(public_point), &session);
  if (result != HEARPORT_SPAKE2_OK || session == nullptr) {
    return std::nullopt;
  }
  return Spake2Session(reinterpret_cast<Opaque*>(session), public_point);
#else
  (void)role;
  (void)scalar;
  (void)identity_a;
  (void)identity_b;
  return std::nullopt;
#endif
}

std::optional<Spake2Result> Spake2Session::Finish(
    std::span<const std::byte> peer_point) && noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (session_ == nullptr || peer_point.size() != kSpake2PointBytes) {
    return std::nullopt;
  }
  auto* session = reinterpret_cast<hearport_spake2_session*>(session_);
  session_ = nullptr;
  hearport_spake2_output* output = nullptr;
  Bytes32 confirmation{};
  const auto result = hearport_spake2_finish(
      session, AsU8(peer_point), peer_point.size(), &output,
      AsU8(confirmation));
  if (result != HEARPORT_SPAKE2_OK || output == nullptr) {
    return std::nullopt;
  }
  std::array<std::byte, kSpake2SessionKeyBytes> session_key{};
  if (hearport_spake2_copy_session_key(output, AsU8(session_key)) !=
      HEARPORT_SPAKE2_OK) {
    hearport_spake2_output_free(output);
    return std::nullopt;
  }
  return Spake2Result(reinterpret_cast<Spake2Result::Opaque*>(output),
                      confirmation, session_key);
#else
  (void)peer_point;
  return std::nullopt;
#endif
}

Spake2Result::~Spake2Result() {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (output_ != nullptr) {
    hearport_spake2_output_free(
        reinterpret_cast<hearport_spake2_output*>(output_));
  }
#endif
}

Spake2Result::Spake2Result(Spake2Result&& other) noexcept
    : output_(other.output_),
      confirmation_(other.confirmation_),
      session_key_(other.session_key_) {
  other.output_ = nullptr;
}

Spake2Result& Spake2Result::operator=(Spake2Result&& other) noexcept {
  if (this == &other) {
    return *this;
  }
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (output_ != nullptr) {
    hearport_spake2_output_free(
        reinterpret_cast<hearport_spake2_output*>(output_));
  }
#endif
  output_ = other.output_;
  confirmation_ = other.confirmation_;
  session_key_ = other.session_key_;
  other.output_ = nullptr;
  return *this;
}

bool Spake2Result::VerifyPeerConfirmation(
    std::span<const std::byte> confirmation) const noexcept {
#if HEARPORT_HAS_SPAKE2_PROVIDER
  if (output_ == nullptr || confirmation.size() != kSha256Bytes) {
    return false;
  }
  return hearport_spake2_verify_confirmation(
             reinterpret_cast<const hearport_spake2_output*>(output_),
             AsU8(confirmation)) == HEARPORT_SPAKE2_OK;
#else
  (void)confirmation;
  return false;
#endif
}

PairingWindow::PairingWindow(std::chrono::seconds duration,
                             std::uint32_t maximum_failures) noexcept
    : duration_(duration), maximum_failures_(maximum_failures) {}

void PairingWindow::Open(TimePoint now) noexcept {
  opened_at_ = now;
  failure_count_ = 0;
  open_ = duration_.count() > 0 && maximum_failures_ > 0;
}

void PairingWindow::Close() noexcept {
  open_ = false;
}

bool PairingWindow::IsOpen(TimePoint now) noexcept {
  if (!open_ || now - opened_at_ >= duration_) {
    open_ = false;
    return false;
  }
  return failure_count_ < maximum_failures_;
}

bool PairingWindow::RecordFailure(TimePoint now) noexcept {
  if (!IsOpen(now)) {
    return false;
  }
  ++failure_count_;
  if (failure_count_ >= maximum_failures_) {
    open_ = false;
    return false;
  }
  return true;
}

}  // namespace hearport::security

File: ios/HearPortReceiver/Sources/HearPortReceiver/Pairing.swift
import Foundation

#if canImport(CryptoKit)
import CryptoKit
#endif

#if canImport(Security)
import Security
#endif

public enum PairingSecurityError: Error, Equatable {
    case invalidPin
    case invalidLength
    case randomGenerationFailed
    case providerUnavailable
    case providerFailure(Int32)
}

public enum Spake2Role: UInt8 {
    case windowsPartyA = 1
    case receiverPartyB = 2
}

public struct RememberedCredential: Equatable {
    public let peerID: Data
    public let pairSecret: Data
    public let windowsSPKISHA256: Data

    public init(peerID: Data, pairSecret: Data, windowsSPKISHA256: Data) throws {
        guard peerID.count == 16, pairSecret.count == 32, windowsSPKISHA256.count == 32 else {
            throw PairingSecurityError.invalidLength
        }
        self.peerID = peerID
        self.pairSecret = pairSecret
        self.windowsSPKISHA256 = windowsSPKISHA256
    }
}

public struct PairingSecurity {
    public static let spake2PointBytes = 65
    public static let spake2SessionKeyBytes = 16

    private static let pinDomain = Data("HearPort-SPAKE2-v1".utf8)
    private static let authDomain = Data("HearPort-Auth-v1".utf8)
    private static let windowsIdentityPrefix = Data("HearPort-Windows-v1".utf8)
    private static let receiverIdentity = Data("HearPort-Receiver-v1".utf8)
    private static let p256OrderMinusOne: [UInt8] = [
        0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00,
        0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
        0xbc, 0xe6, 0xfa, 0xad, 0xa7, 0x17, 0x9e, 0x84,
        0xf3, 0xb9, 0xca, 0xc2, 0xfc, 0x63, 0x25, 0x50
    ]

    public static func validatePIN(_ pin: String) -> Bool {
        let bytes = Array(pin.utf8)
        return bytes.count == 6 && bytes.allSatisfy { $0 >= 48 && $0 <= 57 }
    }

    public static func pinScalar(_ pin: String) throws -> Data {
        guard validatePIN(pin) else { throw PairingSecurityError.invalidPin }
        var message = pinDomain
        message.append(0)
        message.append(contentsOf: pin.utf8)
        #if canImport(CryptoKit)
        let digest = Array(SHA256.hash(data: message))
        #else
        throw PairingSecurityError.providerUnavailable
        #endif
        var scalar = digest
        if scalar.lexicographicallyPrecedes(p256OrderMinusOne) == false {
            subtractBigEndian(&scalar, p256OrderMinusOne)
        }
        addOneBigEndian(&scalar)
        return Data(scalar)
    }

    public static func windowsSPKIHash(_ derSubjectPublicKeyInfo: Data) throws -> Data {
        guard !derSubjectPublicKeyInfo.isEmpty else { throw PairingSecurityError.invalidLength }
        #if canImport(CryptoKit)
        return Data(SHA256.hash(data: derSubjectPublicKeyInfo))
        #else
        throw PairingSecurityError.providerUnavailable
        #endif
    }

    public static func identityA(windowsSPKISHA256: Data) throws -> Data {
        guard windowsSPKISHA256.count == 32 else { throw PairingSecurityError.invalidLength }
        var result = windowsIdentityPrefix
        result.append(0)
        result.append(windowsSPKISHA256)
        return result
    }

    public static func identityB() -> Data {
        receiverIdentity
    }

    public static func defaultSpake2Provider() -> any Spake2Provider {
        #if HEARPORT_SPAKE2_PROVIDER
        return RustSpake2Provider()
        #else
        return UnavailableSpake2Provider()
        #endif
    }

    public static func rememberedAuthMAC(
        pairSecret: Data,
        nonce: Data,
        windowsSPKISHA256: Data
    ) throws -> Data {
        guard pairSecret.count == 32, nonce.count == 32, windowsSPKISHA256.count == 32 else {
            throw PairingSecurityError.invalidLength
        }
        #if canImport(CryptoKit)
        var message = authDomain
        message.append(nonce)
        message.append(windowsSPKISHA256)
        let key = SymmetricKey(data: pairSecret)
        return Data(HMAC<SHA256>.authenticationCode(for: message, using: key))
        #else
        throw PairingSecurityError.providerUnavailable
        #endif
    }

    public static func constantTimeEqual(_ left: Data, _ right: Data) -> Bool {
        guard left.count == right.count else { return false }
        var difference: UInt8 = 0
        for (lhs, rhs) in zip(left, right) {
            difference |= lhs ^ rhs
        }
        return difference == 0
    }

    public static func randomBytes(count: Int) throws -> Data {
        guard count >= 0 else { throw PairingSecurityError.invalidLength }
        if count == 0 { return Data() }
        var result = Data(count: count)
        #if canImport(Security)
        let status = result.withUnsafeMutableBytes { rawBuffer in
            SecRandomCopyBytes(kSecRandomDefault, count, rawBuffer.baseAddress!)
        }
        guard status == errSecSuccess else { throw PairingSecurityError.randomGenerationFailed }
        return result
        #else
        throw PairingSecurityError.providerUnavailable
        #endif
    }

    private static func subtractBigEndian(_ value: inout [UInt8], _ modulus: [UInt8]) {
        var borrow: UInt16 = 0
        for index in stride(from: value.count - 1, through: 0, by: -1) {
            let minuend = UInt16(value[index])
            let subtrahend = UInt16(modulus[index]) + borrow
            value[index] = UInt8(truncatingIfNeeded: minuend &- subtrahend)
            borrow = minuend < subtrahend ? 1 : 0
        }
    }

    private static func addOneBigEndian(_ value: inout [UInt8]) {
        var carry: UInt16 = 1
        for index in stride(from: value.count - 1, through: 0, by: -1) {
            let sum = UInt16(value[index]) + carry
            value[index] = UInt8(truncatingIfNeeded: sum)
            carry = sum >> 8
        }
    }
}

public protocol Spake2Exchange {
    var publicPoint: Data { get }
    func finish(peerPoint: Data) throws -> Spake2Output
}

public protocol Spake2Provider {
    func begin(
        role: Spake2Role,
        scalar: Data,
        identityA: Data,
        identityB: Data
    ) throws -> any Spake2Exchange
}

public struct Spake2Output {
    public let sessionKey: Data
    public let confirmation: Data
    private let verifyPeer: (Data) -> Bool

    public init(sessionKey: Data, confirmation: Data, verifyPeer: @escaping (Data) -> Bool) throws {
        guard sessionKey.count == PairingSecurity.spake2SessionKeyBytes, confirmation.count == 32 else {
            throw PairingSecurityError.invalidLength
        }
        self.sessionKey = sessionKey
        self.confirmation = confirmation
        self.verifyPeer = verifyPeer
    }

    public func verifyPeerConfirmation(_ confirmation: Data) -> Bool {
        verifyPeer(confirmation)
    }
}

public struct UnavailableSpake2Provider: Spake2Provider {
    public init() {}

    public func begin(
        role: Spake2Role,
        scalar: Data,
        identityA: Data,
        identityB: Data
    ) throws -> any Spake2Exchange {
        throw PairingSecurityError.providerUnavailable
    }
}

public final class RememberedAuthSession {
    public let credential: RememberedCredential
    private var pendingNonce: Data?

    public init(credential: RememberedCredential) {
        self.credential = credential
    }

    public func issueChallenge(nonce: Data? = nil) throws -> Data {
        let candidate: Data
        if let nonce {
            candidate = nonce
        } else {
            candidate = try PairingSecurity.randomBytes(count: 32)
        }
        guard candidate.count == 32 else { throw PairingSecurityError.invalidLength }
        pendingNonce = candidate
        return candidate
    }

    public func verify(peerID: Data, mac: Data) -> Bool {
        let nonce = pendingNonce
        pendingNonce = nil
        guard let nonce, PairingSecurity.constantTimeEqual(peerID, credential.peerID) else {
            return false
        }
        guard let expected = try? PairingSecurity.rememberedAuthMAC(
            pairSecret: credential.pairSecret,
            nonce: nonce,
            windowsSPKISHA256: credential.windowsSPKISHA256
        ) else {
            return false
        }
        return PairingSecurity.constantTimeEqual(mac, expected)
    }
}

public final class PairingWindow {
    public let duration: TimeInterval
    public let maximumFailures: Int
    public private(set) var failureCount = 0
    private var openedAt: Date?

    public init(duration: TimeInterval = 120, maximumFailures: Int = 5) {
        self.duration = duration
        self.maximumFailures = maximumFailures
    }

    @discardableResult
    public func open(at now: Date = Date()) -> Bool {
        guard duration > 0, maximumFailures > 0 else { return false }
        openedAt = now
        failureCount = 0
        return true
    }

    public func close() {
        openedAt = nil
    }

    public func isOpen(at now: Date = Date()) -> Bool {
        guard let openedAt, now.timeIntervalSince(openedAt) < duration,
              failureCount < maximumFailures else {
            close()
            return false
        }
        return true
    }

    @discardableResult
    public func recordFailure(at now: Date = Date()) -> Bool {
        guard isOpen(at: now) else { return false }
        failureCount += 1
        if failureCount >= maximumFailures {
            close()
            return false
        }
        return true
    }
}

#if HEARPORT_SPAKE2_PROVIDER
@_silgen_name("hearport_spake2_begin")
private func hearportSpake2Begin(
    _ role: UInt8,
    _ scalar: UnsafePointer<UInt8>,
    _ identityA: UnsafePointer<UInt8>?,
    _ identityALength: Int,
    _ identityB: UnsafePointer<UInt8>?,
    _ identityBLength: Int,
    _ publicPoint: UnsafeMutablePointer<UInt8>,
    _ session: UnsafeMutablePointer<OpaquePointer?>
) -> Int32

@_silgen_name("hearport_spake2_finish")
private func hearportSpake2Finish(
    _ session: OpaquePointer,
    _ peerPoint: UnsafePointer<UInt8>,
    _ peerPointLength: Int,
    _ output: UnsafeMutablePointer<OpaquePointer?>,
    _ confirmation: UnsafeMutablePointer<UInt8>
) -> Int32

@_silgen_name("hearport_spake2_verify_confirmation")
private func hearportSpake2VerifyConfirmation(
    _ output: OpaquePointer,
    _ confirmation: UnsafePointer<UInt8>
) -> Int32

@_silgen_name("hearport_spake2_copy_session_key")
private func hearportSpake2CopySessionKey(
    _ output: OpaquePointer,
    _ sessionKey: UnsafeMutablePointer<UInt8>
) -> Int32

@_silgen_name("hearport_spake2_session_free")
private func hearportSpake2SessionFree(_ session: OpaquePointer)

@_silgen_name("hearport_spake2_output_free")
private func hearportSpake2OutputFree(_ output: OpaquePointer)

public final class RustSpake2Provider: Spake2Provider {
    public init() {}

    public func begin(
        role: Spake2Role,
        scalar: Data,
        identityA: Data,
        identityB: Data
    ) throws -> any Spake2Exchange {
        guard scalar.count == 32 else { throw PairingSecurityError.invalidLength }
        var publicPoint = Data(count: PairingSecurity.spake2PointBytes)
        var session: OpaquePointer?
        let status = scalar.withUnsafeBytes { scalarBuffer in
            identityA.withUnsafeBytes { identityABuffer in
                identityB.withUnsafeBytes { identityBBuffer in
                    publicPoint.withUnsafeMutableBytes { pointBuffer in
                        hearportSpake2Begin(
                            role.rawValue,
                            scalarBuffer.baseAddress!.assumingMemoryBound(to: UInt8.self),
                            identityABuffer.baseAddress?.assumingMemoryBound(to: UInt8.self),
                            identityA.count,
                            identityBBuffer.baseAddress?.assumingMemoryBound(to: UInt8.self),
                            identityB.count,
                            pointBuffer.baseAddress!.assumingMemoryBound(to: UInt8.self),
                            &session
                        )
                    }
                }
            }
        }
        guard status == 0, let session else { throw PairingSecurityError.providerFailure(status) }
        return RustSpake2Exchange(session: session, publicPoint: publicPoint)
    }
}

private final class RustSpake2Exchange: Spake2Exchange {
    private var session: OpaquePointer?
    let publicPoint: Data

    init(session: OpaquePointer, publicPoint: Data) {
        self.session = session
        self.publicPoint = publicPoint
    }

    deinit {
        if let session { hearportSpake2SessionFree(session) }
    }

    func finish(peerPoint: Data) throws -> Spake2Output {
        guard peerPoint.count == PairingSecurity.spake2PointBytes,
              let session else { throw PairingSecurityError.invalidLength }
        self.session = nil
        var output: OpaquePointer?
        var confirmation = Data(count: 32)
        let status = peerPoint.withUnsafeBytes { peerBuffer in
            confirmation.withUnsafeMutableBytes { confirmationBuffer in
                hearportSpake2Finish(
                    session,
                    peerBuffer.baseAddress!.assumingMemoryBound(to: UInt8.self),
                    peerPoint.count,
                    &output,
                    confirmationBuffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
                )
            }
        }
        guard status == 0, let output else { throw PairingSecurityError.providerFailure(status) }
        var sessionKey = Data(count: PairingSecurity.spake2SessionKeyBytes)
        let keyStatus = sessionKey.withUnsafeMutableBytes { keyBuffer in
            hearportSpake2CopySessionKey(
                output,
                keyBuffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
            )
        }
        guard keyStatus == 0 else {
            hearportSpake2OutputFree(output)
            throw PairingSecurityError.providerFailure(keyStatus)
        }
        let rustOutput = try RustSpake2Output(
            output: output,
            sessionKey: sessionKey,
            confirmation: confirmation
        )
        return try Spake2Output(
            sessionKey: rustOutput.sessionKey,
            confirmation: rustOutput.confirmation,
            verifyPeer: { [rustOutput] peerConfirmation in
                rustOutput.verify(peerConfirmation)
            }
        )
    }
}

private final class RustSpake2Output {
    let output: OpaquePointer
    let sessionKey: Data
    let confirmation: Data

    init(output: OpaquePointer, sessionKey: Data, confirmation: Data) throws {
        self.output = output
        self.sessionKey = sessionKey
        self.confirmation = confirmation
    }

    func verify(_ confirmation: Data) -> Bool {
        guard confirmation.count == 32 else { return false }
        return confirmation.withUnsafeBytes { confirmationBuffer in
            hearportSpake2VerifyConfirmation(
                output,
                confirmationBuffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
            ) == 0
        }
    }

    deinit {
        hearportSpake2OutputFree(output)
    }
}
#endif
